File: backend/src/main/java/ma/iorecycling/service/SocieteProprietaireService.java
package ma.iorecycling.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import ma.iorecycling.dto.CreateSocieteProprietaireRequest;
import ma.iorecycling.dto.SocieteProprietaireDTO;
import ma.iorecycling.dto.UpdateSocieteProprietaireRequest;
import ma.iorecycling.entity.SocieteProprietaire;
import ma.iorecycling.mapper.SocieteProprietaireMapper;
import ma.iorecycling.repository.SocieteProprietaireRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service pour la gestion des sociétés propriétaires de camions
 */
@Service
@RequiredArgsConstructor
@Slf4j
@Transactional
public class SocieteProprietaireService {
    
    private final SocieteProprietaireRepository societeProprietaireRepository;
    private final SocieteProprietaireMapper societeProprietaireMapper;
    
    /**
     * Crée une nouvelle société propriétaire
     */
    public SocieteProprietaireDTO createSocieteProprietaire(CreateSocieteProprietaireRequest request) {
        log.info("Création d'une nouvelle société propriétaire : {}", request.getRaisonSociale());
        
        SocieteProprietaire societe = societeProprietaireMapper.toEntity(request);
        SocieteProprietaire savedSociete = societeProprietaireRepository.save(societe);
        
        log.info("Société propriétaire créée avec succès : ID {}", savedSociete.getId());
        return societeProprietaireMapper.toDTO(savedSociete);
    }
    
    /**
     * Récupère toutes les sociétés propriétaires avec pagination
     */
    @Transactional(readOnly = true)
    public Page<SocieteProprietaireDTO> getAllSocietesProprietaires(Pageable pageable) {
        return societeProprietaireRepository.findAll(pageable)
                .map(societeProprietaireMapper::toDTO);
    }
    
    /**
     * Récupère toutes les sociétés propriétaires actives
     */
    @Transactional(readOnly = true)
    public List<SocieteProprietaireDTO> getActiveSocietesProprietaires() {
        return societeProprietaireRepository.findByActifTrue()
                .stream()
                .map(societeProprietaireMapper::toDTO)
                .toList();
    }
    
    /**
     * Récupère une société propriétaire par son ID
     */
    @Transactional(readOnly = true)
    public SocieteProprietaireDTO getSocieteProprietaireById(Long id) {
        SocieteProprietaire societe = societeProprietaireRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Société propriétaire non trouvée avec l'ID : " + id));
        
        return societeProprietaireMapper.toDTO(societe);
    }
    
    /**
     * Met à jour une société propriétaire
     */
    public SocieteProprietaireDTO updateSocieteProprietaire(Long id, UpdateSocieteProprietaireRequest request) {
        log.info("Mise à jour de la société propriétaire ID {}", id);
        
        SocieteProprietaire societe = societeProprietaireRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Société propriétaire non trouvée avec l'ID : " + id));
        
        societeProprietaireMapper.updateEntity(societe, request);
        SocieteProprietaire updatedSociete = societeProprietaireRepository.save(societe);
        
        log.info("Société propriétaire mise à jour avec succès : ID {}", id);
        return societeProprietaireMapper.toDTO(updatedSociete);
    }
    
    /**
     * Supprime une société propriétaire
     */
    public void deleteSocieteProprietaire(Long id) {
        log.info("Suppression de la société propriétaire ID {}", id);
        
        if (!societeProprietaireRepository.existsById(id)) {
            throw new IllegalArgumentException("Société propriétaire non trouvée avec l'ID : " + id);
        }
        
        societeProprietaireRepository.deleteById(id);
        log.info("Société propriétaire supprimée avec succès : ID {}", id);
    }
    
    /**
     * Recherche de sociétés propriétaires
     */
    @Transactional(readOnly = true)
    public Page<SocieteProprietaireDTO> search(String search, Pageable pageable) {
        return societeProprietaireRepository.search(search, pageable)
                .map(societeProprietaireMapper::toDTO);
    }
}

File: frontend/src/app/services/societe-proprietaire.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { SocieteProprietaire, CreateSocieteProprietaireRequest, UpdateSocieteProprietaireRequest, Page } from '../models/societe-proprietaire.model';
import { environment } from '../../environments/environment';

/**
 * Service Angular pour gérer les Sociétés Propriétaires de Camions
 * Consomme les APIs du backend AdminSocieteProprietaireController
 */
@Injectable({
  providedIn: 'root'
})
export class SocieteProprietaireService {
  private apiUrl = `${environment.apiUrl}/admin/societes-proprietaires`;

  constructor(private http: HttpClient) {}

  /**
   * Liste toutes les sociétés propriétaires (paginé)
   */
  getAllSocietesProprietaires(
    page: number = 0, 
    size: number = 20, 
    sort: string = 'raisonSociale,asc',
    search?: string
  ): Observable<Page<SocieteProprietaire>> {
    let params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString())
      .set('sort', sort);

    if (search) {
      params = params.set('search', search);
    }

    return this.http.get<Page<SocieteProprietaire>>(this.apiUrl, { params });
  }

  /**
   * Liste toutes les sociétés propriétaires actives (pour sélection dans formulaire)
   */
  getActiveSocietesProprietaires(): Observable<SocieteProprietaire[]> {
    return this.http.get<SocieteProprietaire[]>(`${this.apiUrl}/actives`);
  }

  /**
   * Récupère une société propriétaire par son ID
   */
  getSocieteProprietaireById(id: number): Observable<SocieteProprietaire> {
    return this.http.get<SocieteProprietaire>(`${this.apiUrl}/${id}`);
  }

  /**
   * Crée une nouvelle société propriétaire
   */
  createSocieteProprietaire(request: CreateSocieteProprietaireRequest): Observable<SocieteProprietaire> {
    return this.http.post<SocieteProprietaire>(this.apiUrl, request);
  }

  /**
   * Met à jour une société propriétaire existante
   */
  updateSocieteProprietaire(id: number, request: UpdateSocieteProprietaireRequest): Observable<SocieteProprietaire> {
    return this.http.put<SocieteProprietaire>(`${this.apiUrl}/${id}`, request);
  }

  /**
   * Supprime une société propriétaire
   */
  deleteSocieteProprietaire(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

